### PR TITLE
Limiting results returned from SearchRuns API

### PR DESCRIPTION
**Pull request by mparkhe**
- 'max_results' argument in SearchRuns PB
- support in backend stores
- client API changes

### CODE REVIEW

**Review by sueann — COMMENTED** [review]
didn't read all the tests super carefully but wanted to leave the one comment

File: mlflow/store/abstract_store.py [hunk 1]
```diff
@@ -197,7 +198,7 @@ def get_metric_history(self, run_uuid, metric_key):
         pass
 
     @abstractmethod
-    def search_runs(self, experiment_ids, search_filter, run_view_type):
+    def search_runs(self, experiment_ids, search_filter, run_view_type, max_results):
```
Comment (smurching) [inline mlflow/store/abstract_store.py]: Can we make this an optional argument (with default of `SEARCH_MAX_RESULTS_DEFAULT`) to avoid breaking the AbstractStore search_runs API?

File: mlflow/store/abstract_store.py [hunk 2]
```diff
@@ -206,6 +207,7 @@ def search_runs(self, experiment_ids, search_filter, run_view_type):
         :param search_filter: :py:class`mlflow.utils.search_utils.SearchFilter` object to encode
             search expression or filter string
         :param run_view_type: ACTIVE, DELETED, or ALL runs
+        :param max_results: Maximum number of runs desired
```
Comment (smurching) [inline mlflow/store/abstract_store.py]: Would be nice to add an analogous doc to `list_run_infos` (state that we'll restrict the number of returned runs), unless we plan to remove that API entirely

File: mlflow/store/sqlalchemy_store.py [hunk 3]
```diff
@@ -399,12 +400,20 @@ def set_tag(self, run_uuid, tag):
             self._check_run_is_active(run)
             session.merge(SqlTag(run_uuid=run_uuid, key=tag.key, value=tag.value))
 
-    def search_runs(self, experiment_ids, search_filter, run_view_type):
+    def search_runs(self, experiment_ids, search_filter, run_view_type, max_results):
+        # TODO: push search query into backend database layer
+        if max_results > SEARCH_MAX_RESULTS_THRESHOLD:
+            raise MlflowException("Search API called with a large max_results ({}). Max threshold "
```
Comment (smurching) [inline mlflow/store/sqlalchemy_store.py]: Wording nit/suggestion: `"Invalid value for request parameter max_results. max_results must be at most {}, but got value {}".format(SEARCH_MAX_RESULTS_THRESHOLD, max_results)`

File: tests/store/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -293,15 +294,16 @@ def test_run_info(self):
             else:
                 self.assertEqual(v, v2)
 
-    def _get_run_configs(self, name='test', experiment_id=None, tags=(), parent_run_id=None):
+    def _get_run_configs(self, name='test', experiment_id=None,
+                         tags=(), parent_run_id=None, start_time=int(time.time())):
```
Comment (smurching) [inline tests/store/test_sqlalchemy_store.py]: Unfortunately in Python setting `start_time=int(time.time())` will cause the default value to be initialized once & reused across function calls, instead of computed with each function call, so I'd undo this change.

**Review by smurching — COMMENTED** [review]
Had some small comments, but otherwise LGTM, thanks @mparkhe!

File: mlflow/protos/service.proto [hunk 5]
```diff
@@ -832,6 +832,9 @@ message SearchRuns {
   // Defaults to only active runs.
   optional ViewType run_view_type = 3 [default = ACTIVE_ONLY];
 
+  // Maximum number of runs desired. Max threshold is 50000
```
[the remark this answers is not in the document]
Comment (mparkhe) [reply mlflow/protos/service.proto]: Whoa! Good catch.

File: tests/store/test_sqlalchemy_store.py (hunk 4 above)
Comment (mparkhe) [reply tests/store/test_sqlalchemy_store.py]: Scary stuff... 

Lemme default this to `None` and call `time.time()` inside the function, if not supplied by caller.

**Comment by mparkhe** [comment]
@smurching, @sueann : addressed your feedback, in case you want to take another look at this.

File: mlflow/store/abstract_store.py [hunk 6]
```diff
@@ -207,7 +208,7 @@ def search_runs(self, experiment_ids, search_filter, run_view_type, max_results)
         :param search_filter: :py:class`mlflow.utils.search_utils.SearchFilter` object to encode
             search expression or filter string
         :param run_view_type: ACTIVE, DELETED, or ALL runs
-        :param max_results: Maximum number of runs desired
+        :param max_results: Maximum number of runs desired. Default: 1000
```
Comment (sueann) [inline mlflow/store/abstract_store.py]: Putting the default here (and in other doc strings) seems brittle since it may change, though it is nice to have it for documentation. Not sure if there is a better way though?

**Review by sueann — COMMENTED** [review]
just one comment/question - lgtm otherwise.

File: mlflow/store/abstract_store.py (hunk 6 above)
Comment (mparkhe) [reply mlflow/store/abstract_store.py]: Good one. We could reference the constant, using `py:const` but it displays the constant name and `mlflow.store` is not documented.

image

Also, looks like default arguments are pre-filled with `1000`. So I could go ahead and remove it from the `:param` line in docs.
